configs/example/gem5_library/x86-mi300x-gpu.py

```diff
@@ -0,0 +1,150 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Script to run a full system GPU simulation.
+
+Usage:
+------
+```
+scons build/VEGA_X86/gem5.opt
+./build/VEGA_X86/gem5.opt
+    configs/example/gem5_library/x86-viper-gpu.py
+    --image <disk image>
+    --kernel <kernel>
+    --app <gpu application>
+```
+
+Example:
+--------
+```
+./build/VEGA_X86/gem5.opt
+    configs/example/gem5_library/x86-viper-gpu.py
+    --image ./gem5-resources/src/x86-ubuntu-gpu-ml/disk-image/x86-ubuntu-gpu-ml
+    --kernel ./gem5-resources/src/x86-ubuntu-gpu-ml/vmlinux-gpu-ml
+    --app ./gem5-resources/src/gpu/square/bin.default/square.default
+```
+"""
+
+import argparse
+
+from gem5.coherence_protocol import CoherenceProtocol
+from gem5.components.devices.gpus.amdgpu import MI300X
+from gem5.components.memory.single_channel import SingleChannelDDR4_2400
+from gem5.components.processors.cpu_types import CPUTypes
+from gem5.components.processors.simple_processor import SimpleProcessor
+from gem5.isas import ISA
+from gem5.prebuilt.viper.board import ViperBoard
+from gem5.prebuilt.viper.cpu_cache_hierarchy import ViperCPUCacheHierarchy
+from gem5.resources.resource import (
+    DiskImageResource,
+    FileResource,
+)
+from gem5.simulate.simulator import Simulator
+from gem5.utils.requires import requires
+
+requires(
+    isa_required=ISA.X86,
+    coherence_protocol_required=CoherenceProtocol.GPU_VIPER,
+)
+
+# Kernel, disk, and applications are obtained locally.
+parser = argparse.ArgumentParser()
+
+parser.add_argument(
+    "--image",
+    type=str,
+    required=True,
+    help="Full path to the gem5-resources x86-ubuntu-gpu-ml disk-image.",
+)
+
+parser.add_argument(
+    "--kernel",
+    type=str,
+    required=True,
+    help="Full path to the gem5-resources vmlinux-gpu-ml kernel.",
+)
+
+parser.add_argument(
+    "--app",
+    type=str,
+    required=True,
+    help="Path to GPU application, python script, or bash script to run",
+)
+
+args = parser.parse_args()
+
+# TODO: This memory size is going to be too small to do anything useful
+memory = SingleChannelDDR4_2400(size="3GiB")
+
+# Note: Only KVM and ATOMIC work due to buggy MOESI_AMD_Base protocol.
+processor = SimpleProcessor(cpu_type=CPUTypes.KVM, isa=ISA.X86, num_cores=4)
+
+for core in processor.cores:
+    if core.is_kvm_core():
+        core.get_simobject().usePerf = False
+
+# The GPU must be created first so we can assign CPU-side DMA ports to the
+# CPU cache hierarchy.
+gpu0 = MI300X()
+
+cache_hierarchy = ViperCPUCacheHierarchy(
+    l1d_size="32KiB",
+    l1d_assoc=8,
+    l1i_size="32KiB",
+    l1i_assoc=8,
+    l2_size="1MiB",
+    l2_assoc=16,
+    l3_size="16MiB",
+    l3_assoc=16,
+    # Add our GPU device's CPU-side DMA ports to this hierarchy
+    device_dmas=gpu0.get_cpu_dma_ports(),
+)
+
+board = ViperBoard(
+    clk_freq="3GHz",
+    processor=processor,
+    memory=memory,
+    cache_hierarchy=cache_hierarchy,
```

Comment:
I suggest adding a (optional) `gpu=` parameter here and connecting it in `connect_things` like everything else.

Reply:
Good suggestion, it's even more clean now.